### ISSUE TITLE
Could fasthttp avoid closing connection when using TimeoutHandler?

### ISSUE THREAD

**Issue by fmaturel** [valyala/fasthttp#846]
Hi, 

We are using `fasthttp` in production with a heavy server load (> 6K QPS).

Using `TimeoutHandler`, we've been adding a timeout around 200ms as our HTTP client expects a very fast response from the server (< 100ms), but we've noticed some `Connection: close` Header on HTTP response our side. This is not what we want, as we use `keep-alive` connections to be as fast as possible and would like to avoid reestablishing the TCP connection to process another HTTP request.

In `server.go` (https://github.com/valyala/fasthttp/blob/master/server.go#L2171), the connection is explicitly closed when hitting a timeout. I understand that this is needed to keep the `Context` clean for a further reuse. 
However, is there any way to avoid closing the connection or should we code a similar function and not use `TimeoutHandler`?

Thanks for your time in this awesome server!

**Comment by erikdubbelboer** [comment]
This code hasn't been touched in years so I'm sure it can be improved. My guess is the connection is closed as the reader (`br`) might already contain bytes from a next request which might mess things up. The comment about `fbr` doesn't make sense to me as that first byte would already have been read.

I guess we could change it to only close the connection if there are bytes in `br`.

If I make a branch with these changes would you be willing to test it before I merge it into master? I feel like this part of the code isn't tested much so it's hard to see what impact changes will have without actually running some real traffic on it.

**Comment by fmaturel** [comment]
Yes, I'd be happy to help. Me and my team will try to add a unit test if possible to cover those cases. Diving into the code will be profitable for us too.

Thanks

**Comment by erikdubbelboer** [comment]
@fmaturel I have created a branch here: https://github.com/valyala/fasthttp/tree/timeout-reuse can you test this in production for you and see if improves things?

**Comment by fmaturel** [comment]
@erikdubbelboer I just tested your branch locally with a heavy load test.

- on `master` branch, I get a lot of these when I lower the timeout duration:

```
HTTP response:
status=408 Request Timeout
headers= 
  Date: Mon, 13 Jul 2020 13:55:04 GMT
  Content-Length: 0
  Connection: close
```

- on `timeout-reuse` branch, I get this:

```
HTTP response:
status=408 Request Timeout
headers= 
  Date: Mon, 13 Jul 2020 13:56:56 GMT
  Content-Length: 0
```

I've seen that you dropped code related to the `br` check, and indeed it did not make much sense to me.
So, It looks good in local environment, and we'll proceed to staging very soon.

Finally, is there any way, when timeout occurs, that we can cancel and clean the remaining `ctx` (which is left in use by the timeout out handler)? Should we reset it and return it to the `sync.Pool` to avoid any leak?

**Comment by erikdubbelboer** [comment]
I have looked into being able to release the `ctx` but this requires some really complicated code to prevent any race condition. I'm not sure if that's worth it. In theory a timeout shouldn't happen very often.

In theory you can also write your own TimeoutHandler that could be much more efficient. Something like:
```go
func Handler(ctx *RequestCtx) {
	responseCh := make(chan someResponeInfo)
	go func() {
		// process...
		responseCh <- ...
	}()

	select {
	case response <-responseCh:
		// write response to ctx...
	case <-time.After(time.Millisecond*100):
		// write timeout response to ctx...
	}
}
```

**Comment by fmaturel** [comment]
As a follow up to this issue, branch https://github.com/valyala/fasthttp/tree/timeout-reuse is in production since end of July and we do not experience any `Connection: close` anymore. Can this branch be merged into next release?

You're right about `TimeoutHandler`, we could write our own version, but then we loose the `concurrency` check in https://github.com/valyala/fasthttp/blob/master/server.go#L424: 
```go
concurrencyCh := ctx.s.concurrencyCh
select {
case concurrencyCh <- struct{}{}:
default:
	ctx.Error(msg, StatusTooManyRequests)
	return
}
```
And that is a good server protection when you handle big volumes.
May we use `GetCurrentConcurrency` to check this on our own `TimeoutHandler` then?

**Comment by erikdubbelboer** [comment]
I have merged the timeout-reuse branch into master and I'll tag a new release tomorrow.

You could use `GetCurrentConcurrency()` but that would have to be actively polled instead of waiting on a channel which of course is different. I'm not sure what is best for your case.

I'm closing this issue as the main connection closing issue has been resolved.
